src/uct/ib/base/ib_md.h

```diff
@@ -383,13 +389,37 @@ static inline uint16_t uct_ib_md_atomic_offset(uint8_t atomic_mr_id)
     return 8 * atomic_mr_id;
 }
 
-
-static inline void uct_ib_memh_init_from_mr(uct_ib_mem_t *memh, struct ibv_mr *mr)
+static inline int
+uct_ib_md_take_primary_mr(uct_ib_md_t *md, uct_ib_mem_t *memh, uint64_t flags)
 {
-    memh->lkey = mr->lkey;
-    memh->rkey = mr->rkey;
+    if (!md->relaxed_order) {
+        return 1;
+    }
+
+    if (flags == UCT_IB_MEM_DEREG) {
+        return !(memh->flags & UCT_IB_MEM_FLAG_RELAXED_ORDERING);
+    }
+
+    if (memh->flags == UCT_IB_MEM_FLAG_ODP) {
+        return 1;
+    }
+
+    if (flags == IBV_ACCESS_REMOTE_ATOMIC) {
+        return 0;
+    }
+
+    return flags & IBV_ACCESS_RELAXED_ORDERING;
```

Comment:
this logic is very confusing and should be refactored: it should be used only when packing rkey and not during init/cleanup